src/navigation/reroute.js

```diff
@@ -125,6 +125,20 @@ export function reroute(pendingPromises = [], eventArguments) {
 		}
 		pendingPromises.forEach(promise => promise.resolve(returnValue));
 
+		try {
+			const appChangeEventName = wasNoOp ? "single-spa:no-app-change": "single-spa:app-change";
+			window.dispatchEvent(new CustomEvent(appChangeEventName));
```

Comment:
Three changes to this logic, which improve the behavior for dispatched events:

1) We fire the events synchronously, and before the reroute promise resolves.
2) We fire the events even if we're going to immediately reroute again (no longer in the `else` statement)
3) We fire the events before we "unlock" the reroute function. What I mean by unlock is that we do not set `appChangeUnderway` to false. What this prevents is anybody doing `window.addEventListener("single-spa:routing-event", () => location.hash = "#")` to cause race conditions and inconsistent behavior